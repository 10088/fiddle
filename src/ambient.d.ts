declare module 'monaco-loader';
declare module 'electron-default-menu';
