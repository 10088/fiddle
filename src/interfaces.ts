export interface StringMap<T> {
  [x: string]: T;
}

export type ElectronVersionState = 'downloading' | 'ready' | 'unknown';

export interface GitHubVersion {
  url: string;
  assets_url: string;
  html_url: string;
  tag_name: string;
  target_commitish: string;
  name: string;
  prerelease: boolean;
  created_at: string;
  published_at: string;
  body: string;
}

export interface ElectronVersion extends GitHubVersion {
  state: ElectronVersionState
}

export interface OutputEntry {
  text: string;
  timestamp: number;
}