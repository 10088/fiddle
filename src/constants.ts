import * as electronModule from 'electron';

const electron = electronModule.remote
  ? electronModule.remote
  : electronModule;

export const USER_DATA_PATH = electron.app.getPath('userData');
export const INDEX_HTML_NAME = 'index.html';
export const MAIN_JS_NAME = 'main.js';
export const RENDERER_JS_NAME = 'renderer.js';
export const PACKAGE_NAME = 'package.json';
export const UNTITLED_NAME = 'untitled-fiddle';

