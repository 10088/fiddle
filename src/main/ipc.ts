import { EventEmitter } from 'events';
import { ipcMain } from 'electron';
import { ipcMainEvents } from '../ipc-events';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
export class IpcMainManager extends EventEmitter {
  constructor() {
    super();

    ipcMainEvents.forEach((name) => {
      ipcMain.on(name, (...args) => this.emit(name, ...args));
    });
  }
}

export const ipcMainManager = new IpcMainManager();
