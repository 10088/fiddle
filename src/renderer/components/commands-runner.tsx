import { Button, IButtonProps } from '@blueprintjs/core';
import { observer } from 'mobx-react';
import * as React from 'react';

import { ElectronVersionState } from '../../interfaces';
import { normalizeVersion } from '../../utils/normalize-version';
import { AppState } from '../state';

export interface RunnerState {
}

export interface RunnerProps {
  appState: AppState;
}

/**
 * The runner component is responsible for actually launching the fiddle
 * with Electron. It also renders the button that does so.
 *
 * @class Runner
 * @extends {React.Component<RunnerProps, RunnerState>}
 */
@observer
export class Runner extends React.Component<RunnerProps, RunnerState> {
  public render() {
    const { versions, version, isRunning } = this.props.appState;

    if (!versions || !version || !versions[normalizeVersion(version)]) {
      return null;
    }

    const state = versions[normalizeVersion(version)].state;
    const props: IButtonProps = {};

    if (state === ElectronVersionState.downloading) {
      props.text = 'Downloading';
      props.disabled = true;
      props.loading = true;
    } else if (state === ElectronVersionState.ready) {
      if (isRunning) {
        props.active = true;
        props.text = 'Stop';
        props.onClick = window.ElectronFiddle.app.runner.stop;
        props.icon = 'stop';
      } else {
        props.text = 'Run';
        props.onClick = window.ElectronFiddle.app.runner.run;
        props.icon = 'play';
      }
    } else {
      return null;
    }

    return <Button {...props} />;
  }
}
