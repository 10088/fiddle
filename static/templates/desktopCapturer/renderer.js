const { desktopCapturer } = require('electron')

// The following example shows how to capture video from
// the screen. It also grabs each window, so you could
// just grab video from a single window.
//
desktopCapturer.getSources({ types: ['window', 'screen'] }, (error, sources) => {
  if (error) throw error
  for (let i = 0; i < sources.length; ++i) {
    console.log(sources[i])

    if (sources[i].name.startsWith('Screen')) {
      navigator.mediaDevices.getUserMedia({
        audio: false,
        video: {
          mandatory: {
            chromeMediaSource: 'desktop',
            chromeMediaSourceId: sources[i].id,
            minWidth: 1280,
            maxWidth: 1280,
            minHeight: 720,
            maxHeight: 720
          }
        }
      })
      .then((stream) => handleStream(stream))
      .catch((e) => handleError(e))
      return
    }
  }
})

function handleStream (stream) {
  const video = document.querySelector('video')
  video.srcObject = stream
  video.onloadedmetadata = (e) => video.play()
}

function handleError (e) {
  console.log(e)
}