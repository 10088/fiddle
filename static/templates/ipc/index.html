<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>IPC Example</title>
</head>
<body>
  <h1>IPC Example</h1>
  <p>Open the developer tools to see the IPC in action.</p>
</body>
</html>
