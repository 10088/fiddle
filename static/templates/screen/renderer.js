const { screen } = require('electron')

const screens = screen.getAllDisplays()
const div = document.querySelector('div')

div.innerText = JSON.stringify(screens);
