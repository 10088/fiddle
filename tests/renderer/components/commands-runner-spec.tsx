import { shallow } from 'enzyme';
import * as React from 'react';

import { Runner } from '../../../src/renderer/components/commands-runner';
import { VersionState } from '../../../src/interfaces';
import { ipcRendererManager } from '../../../src/renderer/ipc';

import { StateMock } from '../../mocks/mocks';

jest.mock('../../../src/renderer/file-manager');
jest.mock('../../../src/renderer/npm');
jest.mock('child_process');
jest.mock('fs-extra');

describe('Runner component', () => {
  let store: StateMock;

  beforeEach(() => {
    ({ state: store } = (window as any).ElectronFiddle.app);
    ipcRendererManager.removeAllListeners();
  });

  describe('renders', () => {
    function expectSnapshotToMatch() {
      const wrapper = shallow(<Runner appState={store as any} />);
      expect(wrapper).toMatchSnapshot();
    }

    it('idle', () => {
      store.currentElectronVersion.state = VersionState.installed;
      expectSnapshotToMatch();
    });

    it('running', () => {
      store.currentElectronVersion.state = VersionState.installed;
      store.isRunning = true;
      expectSnapshotToMatch();
    });

    it('installing modules', () => {
      store.currentElectronVersion.state = VersionState.installed;
      store.isInstallingModules = true;
      expectSnapshotToMatch();
    });

    it('VersionState.downloading', () => {
      store.currentElectronVersion.state = VersionState.downloading;
      store.currentElectronVersion.downloadProgress = 50;
      expectSnapshotToMatch();
    });

    it('VersionState.installing', () => {
      store.currentElectronVersion.state = VersionState.installing;
      expectSnapshotToMatch();
    });

    it('VersionState.ready', () => {
      store.currentElectronVersion.state = VersionState.installed;
      expectSnapshotToMatch();
    });

    it('VersionState.not_downloaded', () => {
      store.currentElectronVersion.state = VersionState.not_downloaded;
      expectSnapshotToMatch();
    });
  });
});
