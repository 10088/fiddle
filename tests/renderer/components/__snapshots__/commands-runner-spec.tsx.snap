// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runner component renders default 1`] = `
<Blueprint3.Button
  icon="play"
  onClick={[MockFunction]}
  text="Run"
/>
`;

exports[`Runner component renders running 1`] = `
<Blueprint3.Button
  active={true}
  icon="stop"
  onClick={[MockFunction]}
  text="Stop"
/>
`;
